tests/federation/transport/test_knocking.py

```diff
@@ -0,0 +1,289 @@
+            (EventTypes.Name, {"content": {"name": "A cool room"}, "state_key": ""},),
+            (
+                EventTypes.RoomAvatar,
+                {
+                    "content": {
+                        "info": {
+                            "h": 398,
+                            "mimetype": "image/jpeg",
+                            "size": 31037,
+                            "w": 394,
+                        },
+                        "url": "mxc://example.org/JWEIFJgwEIhweiWJE",
+                    },
+                    "state_key": "",
+                },
+            ),
+            (
+                EventTypes.RoomEncryption,
+                {"content": {"algorithm": "m.megolm.v1.aes-sha2"}, "state_key": ""},
+            ),
+            (
+                EventTypes.CanonicalAlias,
+                {
+                    "content": {"alias": canonical_alias, "alt_aliases": []},
+                    "state_key": "",
+                },
+            ),
+        ]
+    )
+
+    for event_type, event_dict in room_state.items():
+        event_content, state_key = event_dict.values()
+
+        testcase.get_success(
+            event_injection.inject_event(
+                hs,
+                room_version=KNOCK_UNSTABLE_IDENTIFIER,
+                room_id=room_id,
+                sender=sender,
+                type=event_type,
+                state_key=state_key,
+                content=event_content,
+            )
+        )
+
+    return room_state
+
+
+def check_knock_room_state_against_room_state(
+    testcase: HomeserverTestCase,
+    knock_room_state: List[Dict],
+    expected_room_state: Dict,
+) -> None:
+    """Test a list of stripped room state events received over federation against an
+    dict of expected state events.
+
+    Args:
+        testcase: The testcase that is currently active.
+        knock_room_state: The list of room state that was received over federation.
+        expected_room_state: A dict containing the room state we expect to see in
+            `knock_room_state`.
+    """
+    for event in knock_room_state:
+        event_type = event["type"]
+        testcase.assertIn(event_type, expected_room_state)
+
+        # Check the state content matches
+        testcase.assertEquals(
+            expected_room_state[event_type]["content"], event["content"]
+        )
+
+        # Check the state key is correct
+        testcase.assertEqual(
+            expected_room_state[event_type]["state_key"], event["state_key"]
+        )
+
+        # Ensure the event has been stripped
+        testcase.assertNotIn("signatures", event)
+
+        # Pop once we've found and processed a state event
+        expected_room_state.pop(event_type)
+
+    # Check that all expected state events were accounted for
+    testcase.assertEqual(len(expected_room_state), 0)
+
+    # Ensure that no excess state was included
+    testcase.assertNotIn(SECRET_STATE_EVENT_TYPE, knock_room_state)
```

Comment:
Isn't this already done via the `testcase.assertIn(event_type, expected_room_state)` line above?

Reply:
Nope, as we don't expect to see the secret state event in our expected room state :D This is an additional check to make sure it's *not* in the state returned when knocking.

Reply:
But we've iterated all of the returned state and ensured the keys were in our expected state, so don't we know that this wasn't in the returned state? (Since it wasn't in the expected state.)

Reply:
If it is just an additional check, that's fine I just want to make sure I'm understanding the flow here.

Reply:
Ah right, I see where you're coming from. So the simplified flow is:

```
t = state events we want to see from the remote server
s = state events we actually got from the remote server
x = state event that is in the room, but that we do *not* expect to get from the remote server

for event in s:
    check that 'event' is one of the those in t

    t.pop(event)  # note that we're popping out of *t* here

t should be empty (we saw all the event types we wanted to see)

s should *not* have contained x
```

It might be slightly clearer if I move that last check above the for loop?

Reply:
Nevermind! I see what you meant, this is indeed already done by checking that the event type exists in `expected_room_state`. Sorry it took me a second to get there.

I've removed the extra check and added some clarifying comments.